content/topics/sql-schema-design/index.md

```diff
@@ -0,0 +1,143 @@
+>Design a Umuzi's department database schema to handle data related to it.
+
+The strategist on the team has added the following details to your ticket:
+> The database you are about to design should handle the following queries
+-  if the user has a recruit's full name he/she should be able to find which cohort the recruit belongs to.
+-  if the user has a cohort name he/she should be able to find recruit(s) in it.
+-  if the user has a staff member's email he/she should be able to retrieve recruits  he/she is servicing
+
+
+## The steps
+
+  Step 1:   idetify `entities` within a deparment
+  ---
+
+ Our "Umuzi" schema to include the following entities:
+
+ - Departments
+ - Cohorts
+ - Recruits
+ - staff
+
+![step 1 image](./img/step_1.png)
+
+
+  Step 2: Add more attributes
+  ---
+let establish all entities with their names and properties
+
+
+![step 2 image](./img/step_2.png)
+
+  Step 3: establish  relationships
+  ---
+
+  Each entity, attribute and relationship, should have appropriate names and that can be easily understood by the non-technical people as well. A relationship should connect entities.
+
+ our entities within the "department schema" have the following relationships:
+
+- Each department has a one or more staff member, a staff member is part of only one department at a time
+- each cohort is linked to one or more department(s)
+- each cohort has one or more recruits
+
+We will consider only this scenario for now.
+
+![step 3 image](./img/step_3.png)
+
+Step 4: Create queries
+
+So far we have identified entities, added attributed to our entities, and established relationships. It about time we create queries as required on our ticket.
+
+Do you recall the 4 queries? Great!
+
+1. if the user has a recruit's full name he/she should be able to find which cohort the recruit belongs to.
+
+Let have a closer look at the relationship between recruits and cohorts, one recruits should belong to one cohort at a time, `one to one relationship`.
+
+![recruit -> cohort relatioship](./img/recruit_to_cohort.png)
+
+From this relationship a SQL query could look as follow :
+
+```
+SELECT cohorts.name FROM cohorts
+ JOIN recruits ON cohorts.cohort.id=recruits.cohort_id
+WHERE recruits.name ='Lufuno Bourgier'
+```
+
+-  if the user has a cohort name he/she should be able to find recruit(s) in it.
+
+This is a reverse operation from the previous one, `one to many`.
+
+```
+SELECT recruits.name
+ FROM recruits
+ JOIN cohorts ON recruits.cohort_id = cohorts.id
+WHERE cohort.id = 'Cx'
+ ```
+
+-  if the user has a staff member's email he/she should be able to retrieve recruits  he/she is servicing
+
+There seems to be no direct link between a recruits entity and staff entity, however, they both share one property which is a `department id`, once the relationship between staff entity and the recruits entity is established, we will what is called a `many to many` relationships as illustrated bellow.
```

Comment:
This works but there is another way to reference a many to many relationship that is a bit more versatile.

Consider:

a `department` has many `content_items`, and a `content_item` can be in many `departments`. For example the data science curriculum has content items for `git` and `linear regressions`. But `git` is also a part of the web dev curriculum.

Now how do you represent this relationship? 

The standard thing would be to have a seperate table to represent the relationship.

eg:
department_content_items
- id
- department_id
- content_id

This is a nice tidy way to make many to many relationships because the whole truth about the relationship is in one table. The departments table doesn't need to be changed at all.